ValidatingAdmissionPolicy: support namespace access

cici37 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Support namespace access from CEL expression.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added namespace access support to the CEL expressions of ValidatingAdmissionPolicy via a `namespaceObject`
variable with expressions.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3488
```

cici37 committed:
Support namespace access from cel expression in validatingadmissionpolicy.

cici37 commented:
/triage accepted

cici37 committed:
Whitelist the exposed fields in namespace object and add test

benluddy commented:
/cc

jiahuif commented:
/assign

jiahuif commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/matchconditions/matcher_test.go:
```diff
@@ -19,6 +19,7 @@ package matchconditions
 import (
 	"context"
 	"errors"
+	api "k8s.io/api/core/v1"
```
Check with me offline if I may help you setup goimports to prevent this kind of problems once and for all.

jiahuif commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/interface.go:
```diff
@@ -71,6 +72,8 @@ type Matcher interface {
 	// resource request
 	BindingMatches(a admission.Attributes, o admission.ObjectInterfaces, definition *v1alpha1.ValidatingAdmissionPolicyBinding) (bool, error)
+
+	GetNamespace(name string) (*corev1.Namespace, error)
```
Can we doc what happens if the object is cluster-scoped (without a namespace) here?

cici37 replied:
It is documented in https://github.com/kubernetes/kubernetes/blob/093afa03ba551d131b18cca84ae17e125e72ad44/staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go#L330. I will add more comments when I got my access back :)

jiahuif commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/filter.go:
```diff
@@ -286,6 +294,32 @@ func CreateAdmissionRequest(attr admission.Attributes) *admissionv1.AdmissionReq
+	if namespace == nil {
+		return nil
+	}
+	//md := namespace.GetObjectMeta()
```
nit: stray commented out code.

jiahuif commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/filter.go:
```diff
@@ -286,6 +294,32 @@ func CreateAdmissionRequest(attr admission.Attributes) *admissionv1.AdmissionReq
 }
 
+func CreateNamespaceObject(namespace *v1.Namespace) *v1.Namespace {
+	if namespace == nil {
```
This behavior is actually important because that happens for cluster-scoped resource. Let's add the doc for this function before we forget.

jiahuif reviewed (commented):
Overall it's a bit unclear what happens for cluster-scoped objects without a namespace. I did not see a test for this situation either. Otherwise pretty clean and concise.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/compile_test.go:
```diff
@@ -135,6 +140,41 @@ func TestCompileValidatingPolicyExpression(t *testing.T) {
 			envType: environment.NewExpressions,
 		},
+		{
+			name: "valid namespaceObject",
+			expressions: []string{
```
I was also expecting to see tests that check the behavior of this field when the resource is cluster scoped.

jiahuif replied:
I guess E2E test would help that.

jpbetz replied:
Any test that shows that cluster scoped resources are working with this change would be satisfactory to me.

jiahuif replied:
Because of how the tests are implemented right now, the integration test suite tests "what happens when namespace is nil" while the e2e tests can truly test on a cluster-scoped resource.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go:
```diff
@@ -326,6 +326,12 @@ func (c *celAdmissionController) Validate(
 				}
 			}
 
+			namespace, err := c.policyController.matcher.GetNamespace(a.GetNamespace())
+			// if it is cluster scoped, namespace will be null
+			if err != nil {
+				namespace = nil
+			}
```
Can we avoid relying on an error to check if a namespace is cluster scoped?  If `a.GetNamespace()` is empty can we be guaranteed the object is cluster scoped?

I'm concerned that there might be cases where a namespace scoped object results in an error here and we end up mis-handling the case because we blindly interpreted any error to mean that the object was cluster scoped.  I would rather we do a clean "has namespace" check and then report any errors when we call `matcher.GetNamespace()` using normal error handling.

jiahuif replied:
`a.GetNamespace()` returns empty string for cluster-scoped resources. I've added check to see if the namespace name is empty.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/compile_test.go:
```diff
@@ -49,6 +49,11 @@ func TestCompileValidatingPolicyExpression(t *testing.T) {
 			expressions: []string{"object.foo < params.x"},
 			hasParams:   true,
 		},
+		{
```
Please also add integration tests for this change. I would expect at least one integration test for namespace scoped and one for cluster scoped.

jiahuif replied:
not in `compile_test.go` though. The expressions are not evaluated so whether namespace exists does not matter.

jiahuif commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/compile.go:
```diff
@@ -84,6 +85,56 @@ func BuildRequestType() *apiservercel.DeclType {
 }
 
+// BuildNamespaceType generates a DeclType for Namespace.
+// Certain nested fields in Namespace (e.g. managedFields, ownerReferences etc.) are omitted in the generated DeclType
+// by design.
```
If use the schemaResolver that the type checker uses, these SSA fields are already omitted.

jiahuif replied:
I've reconsidered this matter. I think it is not worthwhile to wire in the whole OpenAPI definitions only for one type. I am planning to compromise with a test to ensure this hand-crafted namespace schema matches that of OpenAPI. What do you think?

jiahuif replied:
Let's postponed the check a bit later. The handcrafted type looks consistent with `v1.Namespace`. We can find a way to sync this later in next release since there is no change of behavior on the user's side.

cici37 replied:
> I've reconsidered this matter. I think it is not worthwhile to wire in the whole OpenAPI definitions only for one type. I am planning to compromise with a test to ensure this hand-crafted namespace schema matches that of OpenAPI. What do you think?

SGTM!

jiahuif committed:
better handling of cluster-scoped resources.

jiahuif commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/compile.go:
```diff
@@ -33,6 +33,7 @@ const (
 	OldObjectVarName                 = "oldObject"
 	ParamsVarName                    = "params"
 	RequestVarName                   = "request"
+	NamespaceVarName                 = "namespaceObject"
```
Should we update the API doc to mention it? That would result a comment-only API review. Or we can add the doc in v1beta1 API PRs.

jpbetz replied:
I'm in favor of the comment-only API review. It should be quick.

cici37 commented:
Thanks @jiahuif for helping addressing comments while I lost my access. Would you mind help to add `namespaceObject` info into API documentation here as well: https://github.com/kubernetes/kubernetes/blob/c906e123057bb84cadd5eb8378faaecc58823060/staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go#L241-L248. Thank you!

jiahuif committed:
[API REVIEW] namespaceObject in Expression doc.

jiahuif committed:
compatibility with composition.

jiahuif committed:
generated: ./hack/update-codegen.sh && ./hack/update-openapi-spec.sh

jiahuif commented:
/label api-review
This is only a one-line comment change.

jiahuif committed:
workaround namespace of namespace is unexpectedly set.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go:
```diff
@@ -319,6 +320,23 @@ func (c *celAdmissionController) Validate(
+			var namespace *v1.Namespace
+			namespaceName := a.GetNamespace()
+
+			// Special case, the namespace object has the namespace of itself (maybe a bug).
+			// unset it if the incoming object is a namespace
```
cc @alexzielenski for visibility since https://github.com/kubernetes/kubernetes/pull/119215 is loosely related

jiahuif committed:
basic test coverage for namespaceObject.

jiahuif commented:
/test pull-kubernetes-e2e-kind-alpha-features

jiahuif commented:
/label tide/merge-method-squash

jiahuif commented:
/test pull-kubernetes-e2e-kind-alpha-features

jpbetz commented:
/lgtm
/approve

/assign @deads2k 
For API review of API documentation comment change

jiahuif commented:
Typo fix. ` 'namespaceObject`  ` to `'namespaceObject'`
/test pull-kubernetes-e2e-kind-alpha-features

jiahuif commented:
re-run codegen
/test pull-kubernetes-e2e-kind-alpha-features

deads2k commented:
api change lgtm

/approve

cici37 commented:
/lgtm Thank you!

jiahuif commented:
/lgtm
